src/aere_repl.erl

```diff
@@ -141,6 +142,32 @@ apply_command(State, type, I) ->
     TypeStr = aeso_ast_infer_types:pp_type("", Type),
     TypeStrClean = re:replace(TypeStr, ?TYPE_CONTAINER ++ "[0-9]*\\.", "", [global, {return, list}]),
     {aere_msg:output(TypeStrClean), State};
+apply_command(State = #repl_state{contract_state = {_, OldType, _}}, state, I) ->
+    [Expr] = aere_sophia:parse_body(I),
+    Contract = aere_mock:eval_contract([Expr], State),
+    TAst = aere_sophia:typecheck(Contract, [dont_unfold]),
+    {_, Type} = aere_sophia:type_of(TAst, ?USER_INPUT),
+    {StateVal, _, _} = compile_and_run_contract(Contract, State),
+
+    %% No need for the annotations from the mock contract
+    NewType = aeso_syntax:set_ann(aere_mock:ann(), Type),
+    NewValue = aeso_syntax:set_ann(aere_mock:ann(), Expr),
+
+    NewState0 =
+        case OldType =:= NewType of
```

Comment:
Tbh, since you can always force a new state using `put` without changing type, I wouldn't make this check and always clear the context